typed-protocols/src/Network/TypedProtocol/Driver/Limits.hs

```diff
@@ -0,0 +1,189 @@
+                          | ExceededTimeLimit
+  deriving Show
+
+instance Exception ProtocolLimitFailure
+
+
+driverWithLimits :: forall ps peerid failure bytes m.
+                    (MonadThrow m, MonadTimer m, Exception failure)
+                 => Tracer m (TraceSendRecv ps peerid failure)
+                 -> peerid
+                 -> Codec ps failure m bytes
+                 -> ProtocolLimits ps bytes
+                 -> Channel m bytes
+                 -> Driver ps (Maybe bytes) m
+driverWithLimits tr peerid Codec{encode, decode}
+                 ProtocolLimits{timeLimitForState, sizeLimitForState, dataSize}
+                 channel@Channel{send} =
+    Driver { sendMessage, recvMessage, startDState = Nothing }
+  where
+    sendMessage :: forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+                   PeerHasAgency pr st
+                -> Message ps st st'
+                -> m ()
+    sendMessage stok msg = do
+      traceWith tr (TraceSendMsg peerid (AnyMessage msg))
+      send (encode stok msg)
+
+    recvMessage :: forall (pr :: PeerRole) (st :: ps).
+                   PeerHasAgency pr st
+                -> Maybe bytes
+                -> m (SomeMessage st, Maybe bytes)
+    recvMessage stok trailing = do
+      decoder <- decode stok
+      let sizeLimit = sizeLimitForState stok
+          timeLimit = fromMaybe (-1) (timeLimitForState stok)
+      result  <- timeout timeLimit $
+                   runDecoderWithLimit sizeLimit dataSize
+                                       channel trailing decoder
+      case result of
+        Just (Right x@(SomeMessage msg, _trailing')) -> do
+          traceWith tr (TraceRecvMsg peerid (AnyMessage msg))
+          return x
+        Just (Left (Just failure)) -> throwM failure
+        Just (Left Nothing)        -> throwM ExceededSizeLimit
+        Nothing                    -> throwM ExceededTimeLimit
+
+runDecoderWithLimit
+    :: forall m bytes failure a. Monad m
+    => Word
+    -- ^ message size limit
+    -> (bytes -> Word)
+    -- ^ byte size
+    -> Channel m bytes
+    -> Maybe bytes
+    -> DecodeStep bytes failure m a
+    -> m (Either (Maybe failure) (a, Maybe bytes))
+runDecoderWithLimit limit size Channel{recv} =
+    go 0
+  where
+    -- Our strategy here is as follows...
+    --
+    -- We of course want to enforce the maximum data limit, but we also want to
+    -- detect and report when we exceed the limit rather than having it be
+    -- misclassified as a generic decode error. For example if we simply limited
+    -- the decoder input to the maximum size then the failure would be reported
+    -- as an unexpected end of input, rather than that the size limit was
+    -- exceeded.
+    --
+    -- So our strategy is to allow the last chunk of input to exceed the limit.
+    -- This leaves just one special case: if the decoder finishes with that
+    -- final chunk, we must check if it consumed too much of the final chunk.
```

Comment:
can the remote peer influence how big is the last chunk?

Reply:
That depends on the `recv` which in our case is up to the mux.

But if the peer can force an unbounded chunk then you've already lost, so it's already up to the channel impl to ensure that chunks are bounded. And then once that's the case it doesn't matter if we go over by part of the final chunk.